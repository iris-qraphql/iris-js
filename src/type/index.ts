export type { Path as ResponsePath } from '../jsutils/Path';

export {
  // Predicate
  isSchema,
  // Assertion
  assertSchema,
  // GraphQL Schema definition
  GraphQLSchema,
} from './schema';
export type { GraphQLSchemaConfig, GraphQLSchemaExtensions } from './schema';

export type { GraphQLScalarType } from './definition';
export {
  isType,
  isScalarType,
  isObjectType,
  isUnionType,
  isEnumType,
  isInputObjectType,
  isListType,
  isNonNullType,
  isInputType,
  isOutputType,
  isLeafType,
  isResolverType,
  isAbstractType,
  isWrappingType,
  isNullableType,
  isNamedType,
  isRequiredArgument,
  isRequiredInputField,
  assertScalarType,
  assertObjectType,
  assertResolverType,
  assertListType,
  assertNonNullType,
  assertLeafType,
  assertCompositeType,
  assertAbstractType,
  getNullableType,
  getNamedType,
  GraphQLList,
  GraphQLNonNull,
} from './definition';

export type {
  GraphQLType,
  GraphQLInputType,
  GraphQLOutputType,
  GraphQLLeafType,
  IrisResolverType,
  GraphQLWrappingType,
  GraphQLNullableType,
  GraphQLNamedType,
  GraphQLNamedInputType,
  GraphQLNamedOutputType,
  ThunkObjMap,
  GraphQLArgument,
  GraphQLArgumentConfig,
  IrisDataVariant,
  GraphQLField,
  GraphQLFieldConfig,
  GraphQLFieldConfigArgumentMap,
  GraphQLFieldConfigMap,
  GraphQLFieldMap,
  GraphQLFieldResolver,
  GraphQLInputField,
  GraphQLIsTypeOfFn,
  GraphQLResolveInfo,
  GraphQLTypeResolver,
  GraphQLScalarSerializer,
  GraphQLScalarValueParser,
  GraphQLScalarLiteralParser,
} from './definition';

export {
  // Predicate
  isDirective,
  // Assertion
  assertDirective,
  // Directives Definition
  GraphQLDirective,
  // Built-in Directives defined by the Spec
  isSpecifiedDirective,
  specifiedDirectives,
  GraphQLIncludeDirective,
  GraphQLSkipDirective,
  GraphQLDeprecatedDirective,
  GraphQLSpecifiedByDirective,
  // Constant Deprecation Reason
  DEFAULT_DEPRECATION_REASON,
} from './directives';

export type {
  GraphQLDirectiveConfig,
  GraphQLDirectiveExtensions,
} from './directives';

// Common built-in scalar instances.
export {
  // Predicate
  isSpecifiedScalarType,
  // Standard GraphQL Scalars
  specifiedScalarTypes,
  GraphQLInt,
  GraphQLFloat,
  GraphQLString,
  GraphQLBoolean,
  GraphQLID,
  // Int boundaries constants
  GRAPHQL_MAX_INT,
  GRAPHQL_MIN_INT,
} from './scalars';

export {
  // Predicate
  isIntrospectionType,
  // GraphQL Types for introspection.
  introspectionTypes,
  __Schema,
  __Directive,
  __DirectiveLocation,
  __Type,
  __Field,
  __InputValue,
  __EnumValue,
  __TypeKind,
  // "Enum" of Type Kinds
  TypeKind,
  // Meta-field definitions.
  SchemaMetaFieldDef,
  TypeMetaFieldDef,
  TypeNameMetaFieldDef,
} from './introspection';

// Validate GraphQL schema.
export { validateSchema, assertValidSchema } from './validate';

// Upholds the spec rules about naming.
export { assertName, assertEnumValueName } from './assertName';
