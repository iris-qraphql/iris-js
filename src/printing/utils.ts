import type { Maybe } from '../utils/type-level';

/**
 * Given maybeArray, print an empty string if it is null or empty, otherwise
 * print all items together separated by separator if provided
 */
export function join(
  maybeArray: Maybe<ReadonlyArray<string | undefined>>,
  separator = '',
): string {
  return maybeArray?.filter((x) => x).join(separator) ?? '';
}

/**
 * Given array, print each item on its own line, wrapped in an indented `{ }` block.
 */
export const block = (array: ReadonlyArray<string | undefined>) =>
  array.length > 0 ? `{\n${indent(join(array, '\n'))}\n}` : '{}';

/**
 * If maybeString is not null or empty, then wrap with start and end, otherwise print an empty string.
 */
export function wrap(
  start: string,
  maybeString: Maybe<string>,
  end: string = '',
): string {
  return maybeString != null && maybeString !== ''
    ? start + maybeString + end
    : '';
}

export function indent(str: string): string {
  return wrap('  ', str.replace(/\n/g, '\n  '));
}
