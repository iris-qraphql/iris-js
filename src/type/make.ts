import type { GraphQLScalarTypeConfig } from 'graphql';
import { GraphQLScalarType } from 'graphql';

import type { IrisType } from './definition';
import { IrisTypeDefinition, IrisTypeRef } from './definition';

export const emptyDataType = (name: string) =>
  new IrisTypeDefinition({ role: 'data', name, variants: [] });

export const gqlScalar = <I, O>(config: GraphQLScalarTypeConfig<I, O>) =>
  new IrisTypeDefinition({
    role: 'data',
    name: config.name,
    description: config.description,
    scalar: new GraphQLScalarType(config),
    variants: [],
  });

export const maybe = <T extends IrisType>(ofType: T) =>
  new IrisTypeRef('MAYBE', ofType);

export const gqlList = <T extends IrisType>(ofType: T) =>
  new IrisTypeRef('LIST', ofType);
