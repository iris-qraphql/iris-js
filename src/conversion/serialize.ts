import { isNil } from 'ramda';

import type {
  IrisDataType,
  IrisStrictType,
  IrisVariant,
} from '../type/definition';
import { isMaybeType, isTypeRef } from '../type/definition';

import { irisError } from '../error';
import { inspect } from '../utils/legacy';
import { isIterableObject, isObjectLike } from '../utils/ObjMap';
import type { IrisMaybe, Maybe } from '../utils/type-level';

type JSON = unknown;

type Serializer<T> = (value: unknown, type: T) => Maybe<JSON>;

export const serializeValue: Serializer<IrisStrictType> = (value, type) => {
  if (isTypeRef(type)) {
    const itemType = type.ofType;

    switch (type.kind) {
      case 'MAYBE':
        return isNil(value) ? null : serializeValue(value, itemType);
      case 'LIST': {
        return serializeList(value, itemType);
      }
    }
  }

  return parseDataType(value, type);
};

const serializeList: Serializer<IrisStrictType> = (value, type) => {
  if (!isIterableObject(value)) {
    throw new TypeError('Required Type!');
  }

  const valuesNodes = [];
  for (const item of value) {
    valuesNodes.push(serializeValue(item, type));
  }

  return valuesNodes;
};

type IrisObject = {
  name?: string;
  fields: Record<string, unknown>;
};

const parseDataType: Serializer<IrisDataType> = (value, type) => {
  if (isNil(value)) {
    throw new TypeError('Required Type!');
  }

  if (type.isPrimitive) {
    const serialized = type.serialize(value);

    if (typeof serialized === 'number' && !Number.isFinite(serialized)) {
      throw new TypeError(
        `Cannot convert value to AST: ${inspect(serialized)}.`,
      );
    }

    return serialized;
  }

  return parseDataVariant(value, type);
};

const parseDataVariant: Serializer<IrisDataType> = (value, type) => {
  const object = decodeValue(value, type.name);
  const variant = type.variantBy(object.name);
  const variantType = variant.type ? variant.type.variantBy() : variant;
  return parseVariantValue(object, variantType);
};

const decodeValue = (value: unknown, typeName: string): IrisObject => {
  if (typeof value === 'string') {
    return { name: value, fields: {} };
  }

  if (isObjectLike(value) && !isIterableObject(value)) {
    const { __typename, ...fields } = value;
    return { name: __typename as string, fields };
  }

  throw irisError(
    `Data "${typeName}" cannot represent value: ${inspect(value)}.`,
  );
};

const parseVariantValue = (
  { name: __typename, fields }: IrisObject,
  variant: IrisVariant<'data'>,
): IrisMaybe<JSON> => {
  const variantFields = Object.values(variant.fields ?? {});

  if (
    __typename &&
    variantFields.filter(({ type }) => !isMaybeType(type)).length === 0 &&
    Object.values(fields).length === 0
  ) {
    return __typename;
  }

  const fieldNodes: Record<string, JSON> = __typename ? { __typename } : {};

  for (const { name, type } of variantFields) {
    fieldNodes[name] = serializeValue(fields[name], type);
  }

  return fieldNodes;
};
