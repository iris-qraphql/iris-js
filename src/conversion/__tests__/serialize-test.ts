import { buildSchema } from '../../type/buildASTSchema';
import { assertDataType } from '../../type/definition';
import { gqlList, gqlScalar, maybe } from '../../type/make';
import {
  IrisBool,
  IrisFloat,
  IrisID,
  IrisInt,
  IrisString,
} from '../../type/scalars';

import { toJSONError } from '../../utils/toJSONDeep';

import { serializeValue } from '../serialize';

const maybeBool = maybe(IrisBool);

describe('serializeValue', () => {
  it('converts boolean values to ASTs', () => {
    expect(serializeValue(true, maybeBool)).toEqual(true);
    expect(serializeValue(false, maybeBool)).toEqual(false);
    expect(serializeValue(undefined, maybeBool)).toEqual(null);
    expect(serializeValue(null, maybeBool)).toEqual(null);
    expect(serializeValue(0, IrisBool)).toEqual(false);
    expect(serializeValue(1, IrisBool)).toEqual(true);
  });

  it('converts Int values to Int ASTs', () => {
    expect(serializeValue(-1, IrisInt)).toEqual(-1);
    expect(serializeValue(123.0, IrisInt)).toEqual(123);
    expect(serializeValue(1e4, IrisInt)).toEqual(10000);
    expect(() => serializeValue(123.5, IrisInt)).toThrow(
      'Int cannot represent non-integer value: 123.5',
    );

    expect(() => serializeValue(1e40, IrisInt)).toThrow(
      'Int cannot represent non 32-bit signed integer value: 1e+40',
    );

    expect(() => serializeValue(NaN, IrisInt)).toThrow(
      'Int cannot represent non-integer value: NaN',
    );
  });

  it('converts Float values to Int/Float ASTs', () => {
    expect(serializeValue(-1, IrisFloat)).toEqual(-1);

    expect(serializeValue(123.0, IrisFloat)).toEqual(123);

    expect(serializeValue(123.5, IrisFloat)).toEqual(123.5);

    expect(serializeValue(1e4, IrisFloat)).toEqual(10000);

    expect(serializeValue(1e40, IrisFloat)).toEqual(1e40);
  });

  it('converts String values to String ASTs', () => {
    expect(serializeValue('hello', IrisString)).toEqual({
      kind: 'StringValue',
      value: 'hello',
    });

    expect(serializeValue('VALUE', IrisString)).toEqual({
      kind: 'StringValue',
      value: 'VALUE',
    });

    expect(serializeValue('VA\nLUE', IrisString)).toEqual({
      kind: 'StringValue',
      value: 'VA\nLUE',
    });

    expect(serializeValue(123, IrisString)).toEqual({
      kind: 'StringValue',
      value: '123',
    });

    expect(serializeValue(false, IrisString)).toEqual({
      kind: 'StringValue',
      value: 'false',
    });

    expect(serializeValue(null, maybe(IrisString))).toEqual({
      kind: 'NullValue',
    });

    expect(serializeValue(undefined, IrisString)).toEqual(null);
  });

  it('converts ID values to Int/String ASTs', () => {
    expect(serializeValue('hello', IrisID)).toEqual('hello');

    expect(serializeValue('VALUE', IrisID)).toEqual('VALUE');

    expect(serializeValue('VA\nLUE', IrisID)).toEqual('VA\nLUE');

    expect(serializeValue(-1, IrisID)).toEqual('-1');

    expect(serializeValue(123, IrisID)).toEqual('123');

    expect(serializeValue('123', IrisID)).toEqual('123');

    expect(serializeValue('01', IrisID)).toEqual('01');

    expect(() => serializeValue(false, IrisID)).toThrow(
      'ID cannot represent value: false',
    );

    expect(serializeValue(null, maybe(IrisID))).toEqual(null);

    expect(serializeValue(undefined, IrisID)).toEqual(null);
  });

  it('converts using serialize from a custom scalar type', () => {
    const passthroughScalar = gqlScalar({
      name: 'PassthroughScalar',
      serialize(value) {
        return value;
      },
    });

    expect(serializeValue('value', passthroughScalar)).toEqual('value');

    expect(() => serializeValue(NaN, passthroughScalar)).toThrow(
      'Cannot convert value to AST: NaN.',
    );
    expect(() => serializeValue(Infinity, passthroughScalar)).toThrow(
      'Cannot convert value to AST: Infinity.',
    );

    const returnNullScalar = gqlScalar({
      name: 'ReturnNullScalar',
      serialize() {
        return null;
      },
    });

    expect(serializeValue('value', returnNullScalar)).toEqual(null);

    class SomeClass {}

    const returnCustomClassScalar = gqlScalar({
      name: 'ReturnCustomClassScalar',
      serialize() {
        return new SomeClass();
      },
    });

    expect(() => serializeValue('value', returnCustomClassScalar)).toThrow(
      'Cannot convert value to AST: {}.',
    );
  });

  it('does not converts NonNull values to NullValue', () => {
    expect(serializeValue(null, IrisBool)).toEqual(undefined);
  });

  const schema = buildSchema(`
    data MyEnum 
      = HELLO {} 
      | GOODBYE {}

    data MyInputObj = {
      foo: Float?
      bar: MyEnum?
    }

    resolver Query = {
      f: MyInputObj
    }
  `);

  const myEnum = assertDataType(schema.getType('MyEnum'));
  const inputObj = assertDataType(schema.getType('MyInputObj'));

  it('converts string values to Enum ASTs if possible', () => {
    expect(serializeValue('HELLO', myEnum)).toEqual('HELLO');

    // Note: case sensitive
    expect(() => serializeValue('hello', myEnum)).toThrow(
      'Data "MyEnum" cannot represent value: "hello"',
    );

    // Note: Not a valid enum value
    expect(() => serializeValue('UNKNOWN_VALUE', myEnum)).toThrow(
      'Data "MyEnum" cannot represent value: "UNKNOWN_VALUE"',
    );
  });

  it('converts array values to List ASTs', () => {
    expect(serializeValue(['FOO', 'BAR'], gqlList(IrisString))).toEqual([
      'FOO',
      'BAR',
    ]);

    expect(serializeValue(['HELLO', 'GOODBYE'], gqlList(myEnum))).toEqual([
      'HELLO',
      'GOODBYE',
    ]);

    function* listGenerator() {
      yield 1;
      yield 2;
      yield 3;
    }

    expect(serializeValue(listGenerator(), gqlList(IrisInt))).toEqual([
      1, 2, 3,
    ]);
  });

  it('reject invalid lists', () => {
    expect(serializeValue(['FOO', null], gqlList(IrisString))).toEqual(
      undefined,
    );
    expect(serializeValue('FOO', gqlList(IrisString))).toEqual(undefined);
  });

  it('converts data objects', () => {
    const object = { foo: 3, bar: 'HELLO' };
    expect(serializeValue(object, inputObj)).toEqual(object);
  });

  it('converts input objects with explicit nulls', () => {
    expect(serializeValue({ foo: null }, inputObj)).toEqual({ foo: null });
  });

  it('does not converts non-object values as input objects', () => {
    expect(toJSONError(() => serializeValue(5, inputObj))).toEqual({
      message: 'Data "MyInputObj" cannot represent non-string value: 5.',
    });
  });
});

describe('parse simple data variants', () => {
  const schema = buildSchema(`
      data NodeType 
        = Leaf { name: String} 
        | Node { children: [NodeType] }
  `);

  const node = (...children: Array<unknown>) => ({
    __typename: 'Node',
    children,
  });
  const leaf = (name: string) => ({ __typename: 'Leaf', name });

  const nodeType = assertDataType(schema.getType('NodeType'));
  const parseNode = (n: unknown) => serializeValue(n, nodeType);

  it("don't accept non data values", () => {
    expect(() => parseNode(['Leaf'])).toThrowErrorMatchingSnapshot();
    expect(() =>
      parseNode([{ __typename: 'Leaf', name: 'abcd' }]),
    ).toThrowErrorMatchingSnapshot();
  });

  it('require field names for non-empty variants', () => {
    expect(parseNode('Leaf')).toEqual(undefined);
    expect(parseNode({ __typename: 'Leaf' })).toEqual(undefined);
    expect(parseNode({ __typename: 'Node' })).toEqual(undefined);
  });

  it('accept simple leaf', () => {
    expect(parseNode({ __typename: 'Leaf', name: 'abcd' })).toMatchSnapshot();
  });

  it('accept recursive nodes', () => {
    const tree1 = node(leaf('xyz'));
    const tree2 = node(leaf('abc'), tree1);
    const tree3 = node(tree1, tree2);
    const tree4 = node(tree1, node(tree2, tree3));
    const tree5 = node(node(tree1), node(tree2, tree3, tree4));

    expect(parseNode(tree1)).toMatchSnapshot();
    expect(parseNode(tree2)).toMatchSnapshot();
    expect(parseNode(tree3)).toMatchSnapshot();
    expect(parseNode(tree4)).toMatchSnapshot();
    expect(parseNode(tree5)).toMatchSnapshot();
  });

  it('reject invalid recursive nodes', () => {
    const tree1 = node(3);
    const tree2 = node([leaf('abc')], tree1);
    const tree3 = node(tree1, node(tree2));
    const tree4 = node(node(tree1), node(tree2, tree3));
    const tree5 = node(node(tree1), node(tree2, tree3), tree4);

    expect(() => parseNode(tree1)).toThrowErrorMatchingSnapshot();
    expect(() => parseNode(tree2)).toThrowErrorMatchingSnapshot();
    expect(() => parseNode(tree3)).toThrowErrorMatchingSnapshot();
    expect(() => parseNode(tree4)).toThrowErrorMatchingSnapshot();
    expect(() => parseNode(tree5)).toThrowErrorMatchingSnapshot();
  });
});

describe('circular data types', () => {
  const schema = buildSchema(`
      data Node = 
        { children: [NodeType] } 
    
      data NodeType 
        = Leaf { name: String? } 
        | Node

      resolver Query = {
        n: NodeType
      }
  `);

  const node = (...children: Array<unknown>) => ({
    __typename: 'Node',
    children,
  });
  const leaf = (name?: string) => ({ __typename: 'Leaf', name });

  const nodeType = assertDataType(schema.getType('NodeType'));
  const parseNode = (n: unknown) => serializeValue(n, nodeType);

  it('ignore optional fields variants', () => {
    expect(parseNode('Leaf')).toEqual('Leaf');
    expect(parseNode({ __typename: 'Leaf' })).toEqual('Leaf');
    expect(parseNode({ __typename: 'Node' })).toEqual(undefined);
  });

  it('accept simple leaf', () => {
    expect(parseNode({ __typename: 'Leaf', name: 'abcd' })).toMatchSnapshot();
  });

  it('accept circular types', () => {
    const tree1 = node(leaf('xyz'));
    const tree2 = node(leaf('abc'), tree1);
    const tree3 = node(tree1, tree2);
    const tree4 = node(tree1, node(tree2, tree3));
    const tree5 = node(node(tree1), node(tree2, tree3, tree4));

    expect(parseNode(tree1)).toMatchSnapshot();
    expect(parseNode(tree2)).toMatchSnapshot();
    expect(parseNode(tree3)).toMatchSnapshot();
    expect(parseNode(tree4)).toMatchSnapshot();
    expect(parseNode(tree5)).toMatchSnapshot();
  });

  it('reject invalid circular types', () => {
    const tree1 = node(3);
    const tree2 = node([leaf('abc')], tree1);
    const tree3 = node(tree1, node(tree2));
    const tree4 = node(node(tree1), node(tree2, tree3));
    const tree5 = node(node(tree1), node(tree2, tree3), tree4);

    expect(() => parseNode(tree1)).toThrowErrorMatchingSnapshot();
    expect(() => parseNode(tree2)).toThrowErrorMatchingSnapshot();
    expect(() => parseNode(tree3)).toThrowErrorMatchingSnapshot();
    expect(() => parseNode(tree4)).toThrowErrorMatchingSnapshot();
    expect(() => parseNode(tree5)).toThrowErrorMatchingSnapshot();
  });
});
