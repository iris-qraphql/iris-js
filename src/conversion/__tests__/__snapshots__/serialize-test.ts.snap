// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`circular data types accept circular types 1`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Leaf",
      "name": "xyz",
    },
  ],
}
`;

exports[`circular data types accept circular types 2`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Leaf",
      "name": "abc",
    },
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Leaf",
          "name": "xyz",
        },
      ],
    },
  ],
}
`;

exports[`circular data types accept circular types 3`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Leaf",
          "name": "xyz",
        },
      ],
    },
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Leaf",
          "name": "abc",
        },
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Leaf",
              "name": "xyz",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`circular data types accept circular types 4`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Leaf",
          "name": "xyz",
        },
      ],
    },
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Leaf",
              "name": "abc",
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
          ],
        },
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "abc",
                },
                Object {
                  "__typename": "Node",
                  "children": Array [
                    Object {
                      "__typename": "Leaf",
                      "name": "xyz",
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`circular data types accept circular types 5`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Leaf",
              "name": "xyz",
            },
          ],
        },
      ],
    },
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Leaf",
              "name": "abc",
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
          ],
        },
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "abc",
                },
                Object {
                  "__typename": "Node",
                  "children": Array [
                    Object {
                      "__typename": "Leaf",
                      "name": "xyz",
                    },
                  ],
                },
              ],
            },
          ],
        },
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Node",
                  "children": Array [
                    Object {
                      "__typename": "Leaf",
                      "name": "abc",
                    },
                    Object {
                      "__typename": "Node",
                      "children": Array [
                        Object {
                          "__typename": "Leaf",
                          "name": "xyz",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "__typename": "Node",
                  "children": Array [
                    Object {
                      "__typename": "Node",
                      "children": Array [
                        Object {
                          "__typename": "Leaf",
                          "name": "xyz",
                        },
                      ],
                    },
                    Object {
                      "__typename": "Node",
                      "children": Array [
                        Object {
                          "__typename": "Leaf",
                          "name": "abc",
                        },
                        Object {
                          "__typename": "Node",
                          "children": Array [
                            Object {
                              "__typename": "Leaf",
                              "name": "xyz",
                            },
                          ],
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`circular data types accept simple leaf 1`] = `
Object {
  "__typename": "Leaf",
  "name": "abcd",
}
`;

exports[`circular data types must provide required fields on variants 1`] = `"type [NodeType] cannot represent value: undefined."`;

exports[`circular data types reject invalid circular types 1`] = `"type NodeType cannot represent value: 3."`;

exports[`circular data types reject invalid circular types 2`] = `"type NodeType cannot represent value: [{ __typename: \\"Leaf\\", name: \\"abc\\" }]."`;

exports[`circular data types reject invalid circular types 3`] = `"type NodeType cannot represent value: 3."`;

exports[`circular data types reject invalid circular types 4`] = `"type NodeType cannot represent value: 3."`;

exports[`circular data types reject invalid circular types 5`] = `"type NodeType cannot represent value: 3."`;

exports[`parse simple data variants accept recursive nodes 1`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Leaf",
      "name": "xyz",
    },
  ],
}
`;

exports[`parse simple data variants accept recursive nodes 2`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Leaf",
      "name": "abc",
    },
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Leaf",
          "name": "xyz",
        },
      ],
    },
  ],
}
`;

exports[`parse simple data variants accept recursive nodes 3`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Leaf",
          "name": "xyz",
        },
      ],
    },
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Leaf",
          "name": "abc",
        },
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Leaf",
              "name": "xyz",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parse simple data variants accept recursive nodes 4`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Leaf",
          "name": "xyz",
        },
      ],
    },
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Leaf",
              "name": "abc",
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
          ],
        },
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "abc",
                },
                Object {
                  "__typename": "Node",
                  "children": Array [
                    Object {
                      "__typename": "Leaf",
                      "name": "xyz",
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parse simple data variants accept recursive nodes 5`] = `
Object {
  "__typename": "Node",
  "children": Array [
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Leaf",
              "name": "xyz",
            },
          ],
        },
      ],
    },
    Object {
      "__typename": "Node",
      "children": Array [
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Leaf",
              "name": "abc",
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
          ],
        },
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "abc",
                },
                Object {
                  "__typename": "Node",
                  "children": Array [
                    Object {
                      "__typename": "Leaf",
                      "name": "xyz",
                    },
                  ],
                },
              ],
            },
          ],
        },
        Object {
          "__typename": "Node",
          "children": Array [
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Leaf",
                  "name": "xyz",
                },
              ],
            },
            Object {
              "__typename": "Node",
              "children": Array [
                Object {
                  "__typename": "Node",
                  "children": Array [
                    Object {
                      "__typename": "Leaf",
                      "name": "abc",
                    },
                    Object {
                      "__typename": "Node",
                      "children": Array [
                        Object {
                          "__typename": "Leaf",
                          "name": "xyz",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "__typename": "Node",
                  "children": Array [
                    Object {
                      "__typename": "Node",
                      "children": Array [
                        Object {
                          "__typename": "Leaf",
                          "name": "xyz",
                        },
                      ],
                    },
                    Object {
                      "__typename": "Node",
                      "children": Array [
                        Object {
                          "__typename": "Leaf",
                          "name": "abc",
                        },
                        Object {
                          "__typename": "Node",
                          "children": Array [
                            Object {
                              "__typename": "Leaf",
                              "name": "xyz",
                            },
                          ],
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`parse simple data variants accept simple leaf 1`] = `
Object {
  "__typename": "Leaf",
  "name": "abcd",
}
`;

exports[`parse simple data variants don't accept non data values 1`] = `"type NodeType cannot represent value: [\\"Leaf\\"]."`;

exports[`parse simple data variants don't accept non data values 2`] = `"type NodeType cannot represent value: [{ __typename: \\"Leaf\\", name: \\"abcd\\" }]."`;

exports[`parse simple data variants reject invalid recursive nodes 1`] = `"type NodeType cannot represent value: 3."`;

exports[`parse simple data variants reject invalid recursive nodes 2`] = `"type NodeType cannot represent value: [{ __typename: \\"Leaf\\", name: \\"abc\\" }]."`;

exports[`parse simple data variants reject invalid recursive nodes 3`] = `"type NodeType cannot represent value: 3."`;

exports[`parse simple data variants reject invalid recursive nodes 4`] = `"type NodeType cannot represent value: 3."`;

exports[`parse simple data variants reject invalid recursive nodes 5`] = `"type NodeType cannot represent value: 3."`;

exports[`parse simple data variants require field names for non-empty variants 1`] = `"type String cannot represent value: undefined."`;

exports[`parse simple data variants require field names for non-empty variants 2`] = `"type String cannot represent value: undefined."`;

exports[`parse simple data variants require field names for non-empty variants 3`] = `"type [NodeType] cannot represent value: undefined."`;

exports[`serializeValue converts ID values to Int/String ASTs 1`] = `"ID cannot represent value: false"`;

exports[`serializeValue converts ID values to Int/String ASTs 2`] = `"type ID cannot represent value: undefined."`;

exports[`serializeValue converts ID values to Int/String ASTs 3`] = `"type ID cannot represent value: null."`;

exports[`serializeValue converts String values to String ASTs 1`] = `"type String cannot represent value: undefined."`;

exports[`serializeValue converts String values to String ASTs 2`] = `"type String cannot represent value: null."`;

exports[`serializeValue converts using serialize from a custom scalar type 1`] = `"type PassthroughScalar cannot represent value: NaN."`;

exports[`serializeValue converts using serialize from a custom scalar type 2`] = `"type PassthroughScalar cannot represent value: Infinity."`;

exports[`serializeValue does not converts NonNull values to NullValue 1`] = `"type Boolean cannot represent value: null."`;

exports[`serializeValue does not converts non-object values as input objects 1`] = `"type MyInputObj cannot represent value: 5."`;

exports[`serializeValue reject invalid lists 1`] = `"type String cannot represent value: null."`;

exports[`serializeValue reject invalid lists 2`] = `"type [String] cannot represent value: \\"FOO\\"."`;
